### PR TITLE
Add project code

### PR DESCRIPTION
**Pull request by corest**
Towards https://github.com/giantswarm/giantswarm/issues/11373

So this expected to be init container for apps which require vault access.

E.g. tested this on`ginger`/`cluster-service`

```
      initContainers:
      - args:
        - --vault-address=https://vault.ginger.eu-west-1.aws.gigantic.io
        - --vault-policy=cluster-service
        - --vault-token-secret-name=cluster-service-vault-token
        - --vault-token-secret-namespace=giantswarm
        image: quay.io/giantswarm/k8s-jwt-to-vault-token:0.0.0-a79a157338b5f3373d37c62af7ba3bc99a8ee31f
        imagePullPolicy: Always
        name: ensure-vault-token
```

App reads service account jwt, logs with it into vault, write vault token into secret.

Then `cluster-service` deployment references that vault token via env variable

```
      containers:
      - args:
        - daemon
        - --config.dirs=/var/run/cluster-service/configmap/
        - --config.dirs=/var/run/cluster-service/secret/
        - --config.files=config
        - --config.files=secret
        - --service.vault.token=$(VAULT_TOKEN)
        env:
        - name: VAULT_TOKEN
          valueFrom:
            secretKeyRef:
              key: token
              name: cluster-service-vault-token
        image: quay.io/giantswarm/cluster-service:1.2.3
```

With this we get vault token regenerated on every run

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,5 +11,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Add github automation workflows.
 - Add `Dockerfile`.
+- Add code.
```
Comment (kopiczko) [inline CHANGELOG.md]: I think everything before first release should be written in the readme.

```suggestion
- First release.
```

File: cmd/runner.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	kubernetes "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/giantswarm/k8s-jwt-to-vault-token/internal/label"
+	"github.com/giantswarm/k8s-jwt-to-vault-token/pkg/project"
+)
+
+const (
+	jwtFilePath       = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	vaultAuthEndpoint = "v1/auth/kubernetes/login"
```
Comment (kopiczko) [inline cmd/runner.go]: I'd inline values for those consts. They are used in a single place.

File: cmd/runner.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	return nil
+}
+
+func (r *runner) run(ctx context.Context, cmd *cobra.Command, args []string) error {
+	jwt, err := readJWTFromFile(jwtFilePath)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	vaultToken, err := vaultLogin(jwt, r.flag.VaultPolicy, r.flag.VaultAddr)
```
Comment (kopiczko) [inline cmd/runner.go]: Nit: The naming is inconsistent. Is it policy or role?

File: cmd/runner.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonValues))
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // nolint
```
Comment (kopiczko) [inline cmd/runner.go:155]: No way around that?

File: cmd/runner.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	}
+	client := &http.Client{Transport: tr}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
```
Comment (kopiczko) [inline cmd/runner.go:162]: Resp body should be closed here.

File: cmd/runner.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return "", microerror.Maskf(executionFailedError, "expected code %#q got %#q", http.StatusOK, resp.StatusCode)
```
Comment (kopiczko) [inline cmd/runner.go]: Should we print response body here as well for easier debugging?

File: cmd/runner.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	err = json.Unmarshal(body, &tokenData)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
```
Comment (kopiczko) [inline cmd/runner.go:189]: Not sure how confident you are but I have memories of vault API changing so maybe it makes sense to assert the token isn't empty here.

**Comment by kopiczko** [comment]
I think it would be good to put the information from the PR description in the readme.

File: cmd/runner.go (hunk 4 above)
Comment (corest) [reply cmd/runner.go:155]: We would need to pass vault CA into template. But we don't have it in config. Also we've used it in the same way via vault client. It is just hidden there so we never cared about that

**Review by kopiczko — APPROVED** [review]

**Pull request merged**